### ISSUE TITLE
deleteMany middleware on child schema (using discriminator)

### ISSUE THREAD

**Issue by ep2016** [Automattic/mongoose#9561]
I've noticed the following and would like to ask whether this is a *bug* or a *feature* that I need to cope with and use an workaround, which I would appreciate if you could suggest.

**What is the current behavior?**
I am using `discriminator` to define child schemas for a User model and have taken advantage of `pre('deleteMany', async function (next){ next(); });` on one of the child models. When calling `deleteMany` on that child model, the hook never fires. On the contrary, if I move the hook from child to parent schema, it is fired despite the fact that `deleteMany` was called on the child model.

**What is the expected behavior?**
I would expect that the hook would be called when attached to the child schema when `deleteMany` is called on the child model. However, it may have been designed otherwise and I lack the relative know-how. If this is the case, I would be grateful if somebody could explain and suggest how to deal with that. The straightforward workaround is to take advantage of the parent model hook, but I would like to avoid that in order to keep code cleaner and more to the point, since the cascade I want to achieve is meaningful only for one of the child models.

**What are the versions of Node.js, Mongoose and MongoDB you are using? Note that "latest" is not a version.**
Mongoose Version 5.10.15
Node Version v14.14.0

**Comment by vkarpov15** [comment]
The below script shows that the discriminator schema's `pre('deleteMany')` hook is called. Please modify the below script to demonstrate your issue.

```javascript
'use strict';

const mongoose = require('mongoose');

const { Schema } = mongoose;
mongoose.set('debug', true);

run().catch(err => console.log(err));

async function run() {
  await mongoose.connect('mongodb://localhost:27017/test', {
    useNewUrlParser: true,
    useUnifiedTopology: true
  });

  await mongoose.connection.dropDatabase();

  const schema = new Schema({ name: String });
  const Base = mongoose.model('Base', schema);

  const personSchema = Schema({ age: Number });
  personSchema.pre('deleteMany', () => console.log('Discriminator deleteMany!'));
  const Person = Base.discriminator('Person', personSchema);

  await Person.deleteMany({}); // Prints "Discriminator deleteMany!"
  console.log('Done');
}
```

**Comment by ep2016** [comment]
Thanks a lot for the instant response Valeri! 
I think I got the glitch in my code! I probably had defined the `discriminator` **before** defining the pre hook as the modified script below indicates. I changed that after having both posted the issue and totally changed the code. Therefore I did not realize the change. Apologies for that!

`'use strict';

const mongoose = require('mongoose');

const { Schema } = mongoose;
mongoose.set('debug', true);

run().catch(err => console.log(err));

async function run() {
  await mongoose.connect('mongodb://localhost:27017/test', {
    useNewUrlParser: true,
    useUnifiedTopology: true
  });

  await mongoose.connection.dropDatabase();

  const schema = new Schema({ name: String });
  const Base = mongoose.model('Base', schema);

  const personSchema = Schema({ age: Number });
  const Person = Base.discriminator('Person', personSchema);
  personSchema.pre('deleteMany', () => console.log('Discriminator deleteMany!'));

  await Person.deleteMany({}); // Prints "Discriminator deleteMany!"
  console.log('Done');
}`

I would like to ask you **however** how could somebody deal with that `pre('deleteMany'` hook, since it is Query based according to the Docs and therefore you do not have the affected documents in a var like `this`. I've struggled a lot but could not find a way through. I ended up with the following (probably ugly) workaround:

`'use strict';

const mongoose = require('mongoose');

const { Schema } = mongoose;
mongoose.set('debug', true);

run().catch(err => console.log(err));

async function run() {
  await mongoose.connect('mongodb://localhost:27017/test', {
    useNewUrlParser: true,
    useUnifiedTopology: true
  });

  await mongoose.connection.dropDatabase();

  const schema = new Schema({ name: String });
  const Base = mongoose.model('Base', schema);

  const taskSchema = new Schema({ descr: { type: String }, owner: { type: mongoose.Schema.Types.ObjectId, ref: 'Person' } });
  const Task = mongoose.model('Task', taskSchema);

  const personSchema = Schema({ age: Number });
  //personSchema.pre('deleteMany', () => console.log('Discriminator deleteMany!'));
  personSchema.statics.deleteManyWithMiddleware = async () => {
    const collection = await Person.find();
    for (const item of collection) {
        await item.remove();
    }
  };
  personSchema.pre('remove', async function (next) {
    const person = this;
    const tasks = await Task.find({ owner: person._id });
    for (const item of tasks) {
        await item.remove();
    }
    next();
  });
  const Person = Base.discriminator('Person', personSchema);

  await Person.deleteMany({}); // Prints "Discriminator deleteMany!"
  console.log('Done');
}`
